### PR TITLE
Feature metadata handling

### PR DESCRIPTION
**Pull request by Antuan-AE**
To speed up the master metadata loading process, all metadata sections are loaded in parallel. 
Support for metadata versions before v2.9 is deprecated. 
Metadata File with v2.9 splits its content into sections which allows it to be handled in parallel. 
New algorithm is fully tested with two added functional tests to SanityChecks.

### CODE REVIEW

**Review by lgsilva3087 — CHANGES_REQUESTED** [review]
Almost ready. Please see my suggestions.

File: src/master/metadata_loader.cc [hunk 1]
```diff
@@ -0,0 +1,37 @@
+#include "common/platform.h"
```
[the remark this answers is not in the document]
Comment (Antuan-AE) [reply src/master/metadata_loader.cc:1]: done

File: src/master/filesystem_store.cc [hunk 2]
```diff
@@ -773,60 +750,59 @@ int fs_checknodes(int ignoreflag) {
 	return 1;
 }
 
-int fs_loadnodes(FILE *fd) {
-	int s;
-	fs_loadnode(NULL);
+bool fs_loadnodes(MetadataLoader::Options options) {
+	int8_t s;
```
[the remark this answers is not in the document]
Comment (Antuan-AE) [reply src/master/filesystem_store.cc:754]: done

File: src/master/filesystem_store.cc [hunk 3]
```diff
@@ -19,42 +19,48 @@
...
+
+// TODO (Baldor): Review the need for these constants below
 constexpr uint8_t kMetadataVersionLegacy = 0x15;
 constexpr uint8_t kMetadataVersionSaunaFS = 0x16;
 constexpr uint8_t kMetadataVersionWithSections = 0x20;
 constexpr uint8_t kMetadataVersionWithLockIds = 0x29;
-
-char const MetadataStructureReadErrorMsg[] = "error reading metadata (structure)";
+static constexpr int8_t kOpSuccess = 0;
+static constexpr int8_t kOpFailure = -1;
```
[the remark this answers is not in the document]
Comment (Antuan-AE) [reply src/master/filesystem_store.cc:61]: done

File: src/master/filesystem_store.cc [hunk 4]
```diff
@@ -19,42 +19,48 @@
...
 #include "master/matomlserv.h"
 #include "master/metadata_dumper.h"
+#include "master/metadata_loader.h"
 #include "master/restore.h"
 
+#include "master/filesystem_store.h"
+
+// TODO (Baldor): Review the need for these constants below
 constexpr uint8_t kMetadataVersionLegacy = 0x15;
 constexpr uint8_t kMetadataVersionSaunaFS = 0x16;
```
[the remark this answers is not in the document]
Comment (Antuan-AE) [reply src/master/filesystem_store.cc:57]: This is part of the dumping algorithm. Will be remove after the new dumping algorithm implementation is finished.

File: src/master/filesystem_store.cc [hunk 5]
```diff
@@ -19,42 +19,48 @@
...
 #include "master/filesystem_store_acl.h"
-#include "master/locks.h"
+#include "master/filesystem_xattr.h"
 #include "master/matoclserv.h"
 #include "master/matomlserv.h"
 #include "master/metadata_dumper.h"
+#include "master/metadata_loader.h"
 #include "master/restore.h"
 
+#include "master/filesystem_store.h"
```
[the remark this answers is not in the document]
Comment (Antuan-AE) [reply src/master/filesystem_store.cc:53]: done

File: src/master/filesystem.h [hunk 6]
```diff
@@ -231,7 +235,8 @@ uint8_t fs_gettrashpath(uint32_t rootinode,uint8_t sesflags,uint32_t inode,std::
 uint8_t fs_getdetachedattr(uint32_t rootinode,uint8_t sesflags,uint32_t inode,Attributes& attr,uint8_t dtype);
 
 // EXTRA
-uint8_t fs_get_dir_stats(const FsContext &context,uint32_t inode,uint32_t *inodes,uint32_t *dirs,uint32_t *files,uint32_t *chunks,uint64_t *length,uint64_t *size,uint64_t *rsize);
+uint8_t fs_get_dir_stats(const FsContext &context, uint32_t inode, uint32_t *inodes, uint32_t *dirs, uint32_t *files,
+						 uint32_t *links, uint32_t *chunks, uint64_t *length, uint64_t *size, uint64_t *rsize);
```
[the remark this answers is not in the document]
Comment (Antuan-AE) [reply src/master/filesystem.h:239]: done

File: src/master/chunks.cc [hunk 7]
```diff
@@ -34,27 +34,29 @@
...
 #include "common/datapack.h"
-#include "common/exceptions.h"
 #include "common/event_loop.h"
+#include "common/exceptions.h"
 #include "common/flat_set.h"
 #include "common/goal.h"
 #include "common/hashfn.h"
-#include "common/saunafs_version.h"
 #include "common/loop_watchdog.h"
+//#include "common/memory_file.h"
```
[the remark this answers is not in the document]
Comment (Antuan-AE) [reply src/master/chunks.cc:50]: done

File: src/common/parallel_helpers.h
[the remark this answers is not in the document]
Comment (Antuan-AE) [reply src/common/parallel_helpers.h:1]: removed

File: src/common/parallel_helpers.cc
[the remark this answers is not in the document]
Comment (Antuan-AE) [reply src/common/parallel_helpers.cc:1]: removed

File: src/master/chunks.cc [hunk 8]
```diff
@@ -34,27 +34,29 @@
...
 #include "common/massert.h"
+#include "common/saunafs_version.h"
 #include "common/slice_traits.h"
 #include "common/small_vector.h"
-#include "master/chunkserver_db.h"
 #include "master/checksum.h"
 #include "master/chunk_goal_counters.h"
+#include "master/chunkserver_db.h"
 #include "master/filesystem.h"
+#include "master/filesystem_store.h"
```
[the remark this answers is not in the document]
Comment (Antuan-AE) [reply src/master/chunks.cc:58]: done

File: src/master/chunks.h [hunk 9]
```diff
@@ -22,15 +22,16 @@
...
+#include <cinttypes>
+#include <cstdio>
 
 #include "common/chunk_part_type.h"
 #include "common/chunk_type_with_address.h"
 #include "common/chunk_with_address_and_label.h"
 #include "common/chunks_availability_state.h"
-#include "protocol/cltoma.h"
 #include "master/checksum.h"
+#include "master/filesystem_store.h"
```
[the remark this answers is not in the document]
Comment (Antuan-AE) [reply src/master/chunks.h:33]: done

File: src/master/filesystem_store.cc [hunk 10]
```diff
@@ -19,42 +19,48 @@
...
 #include "common/rotate_files.h"
+#include "common/saunafs_version.h"
 #include "common/setup.h"
 
+#include "common/memory_mapped_file.h"
 #include "master/changelog.h"
 #include "master/filesystem.h"
-#include "master/filesystem_xattr.h"
+#include "master/filesystem_checksum.h"
+#include "master/filesystem_freenode.h"
```
[the remark this answers is not in the document]
Comment (Antuan-AE) [reply src/master/filesystem_store.cc:41]: done

**Review by ralcolea — COMMENTED** [review]
PR is almost complete. Please see my suggestions.

I also noticed that some commit messages have minor typos. Consider reviewing the following commits messages before merging the branch:

  * a4c2f1c
  * 3999491
  * 9c6e55e
  * b6a2592
  * e778de2
  * c971826 
  * df44a9b

File: src/master/metadata_loader.cc [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+			the Free Software Foundation, version 3.
+
+ SaunaFS is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with SaunaFS. If not, see <http://www.gnu.org/licenses/>.
+*/#include "common/platform.h"
```
[the remark this answers is not in the document]
Comment (Antuan-AE) [reply src/master/metadata_loader.cc:17]: done

File: src/master/metadata_loader.cc [hunk 12]
```diff
@@ -0,0 +1,53 @@
+/*
+Copyright 2023      Leil Storage OÜ
+
+This file is part of SaunaFS.
+
+	 SaunaFS is free software: you can redistribute it and/or modify
```
[the remark this answers is not in the document]
Comment (Antuan-AE) [reply src/master/metadata_loader.cc:6]: done

File: src/master/locks.cc [hunk 13]
```diff
@@ -18,14 +18,16 @@
...
 #include "master/locks.h"
+#include "common/platform.h"
 
-#include <algorithm>
 #include <functional>
 
+#include "common/memory_mapped_file.h"
 #include "common/slogger.h"
 
+#include "master/locks.h"
```
[the remark this answers is not in the document]
Comment (Antuan-AE) [reply src/master/locks.cc:29]: removed

File: src/master/filesystem_store_acl.cc [hunk 14]
```diff
@@ -80,236 +61,239 @@ void fs_store_acls(FILE *fd) {
...
+                int ignoreFlag) {
 	try {
 		// Read size of the entry
 		uint32_t size = 0;
-		buffer.resize(serializedSize(size));
-		if (fread(buffer.data(), 1, buffer.size(), fd) != buffer.size()) {
-			throw Exception(std::string("read error: ") + strerr(errno), SAUNAFS_ERROR_IO);
+		uint32_t sizeofsize = sizeof(size);
+		uint8_t *ptr;
+        try {
```
[the remark this answers is not in the document]
Comment (Antuan-AE) [reply src/master/filesystem_store_acl.cc:244]: done

File: src/master/filesystem_store_acl.cc [hunk 15]
```diff
@@ -80,236 +61,239 @@ void fs_store_acls(FILE *fd) {
...
 		if (s < 0) {
-			return -1;
+			return false;
 		}
 	} while (s == 0);
 
-	return 0;
-}
+	do {
+		s = fs_load_posix_acl(options.metadataFile, options.offset, options.ignoreFlag, true);
```
[the remark this answers is not in the document]
Comment (Antuan-AE) [reply src/master/filesystem_store_acl.cc:228]: done

File: src/master/filesystem_store_acl.cc [hunk 16]
```diff
@@ -80,236 +61,239 @@ void fs_store_acls(FILE *fd) {
...
 		}
 	} while (s == 0);
+	return true;
+}
 
+bool fs_load_posix_acls(MetadataLoader::Options options) {
+	int s;
 	do {
-		s = fs_load_posix_acl(fd, ignoreflag, true);
+		s = fs_load_posix_acl(options.metadataFile, options.offset, options.ignoreFlag, false);
```
[the remark this answers is not in the document]
Comment (Antuan-AE) [reply src/master/filesystem_store_acl.cc:221]: done

File: src/master/filesystem_store_acl.cc [hunk 17]
```diff
@@ -80,236 +61,239 @@ void fs_store_acls(FILE *fd) {
...
 	}
 }
 
-int fs_load_posix_acls(FILE *fd, int ignoreflag) {
+bool fs_load_legacy_acls(MetadataLoader::Options options) {
 	int s;
 
 	do {
-		s = fs_load_posix_acl(fd, ignoreflag, false);
+		s = fs_load_legacy_acl(options.metadataFile, options.offset, options.ignoreFlag);
```
[the remark this answers is not in the document]
Comment (Antuan-AE) [reply src/master/filesystem_store_acl.cc:210]: done

File: src/master/filesystem_store_acl.cc [hunk 18]
```diff
@@ -80,236 +61,239 @@ void fs_store_acls(FILE *fd) {
...
+	static const int8_t kError = -1;
+	static const int8_t kSuccess = 0;
+	static const int8_t kEndReached = 1;
 	try {
 		// Read size of the entry
 		uint32_t size = 0;
-		buffer.resize(serializedSize(size));
-		if (fread(buffer.data(), 1, buffer.size(), fd) != buffer.size()) {
-			throw Exception(std::string("read error: ") + strerr(errno), SAUNAFS_ERROR_IO);
+        uint32_t sizeofsize = sizeof(size);
```
[the remark this answers is not in the document]
Comment (Antuan-AE) [reply src/master/filesystem_store_acl.cc:74]: done

**Review by ralcolea — APPROVED** [review]
I think all the suggestions were addressed. However, I still notice two typos, probably missed. I list them below:
  * 3eb26f9: master: Add counter to trach amount of link nodes
     - Replace trach by track in the commit title.

  * bb068fb: master: Use callbacks to handle metadata loading
     - Not all the sentences have final point.
     - I think there is a typo in the last sentence, Allign -> Align (or something else).
  
My last suggestion is to consider squashing commits for applying code review suggestions. It will help reducing the number of commits in the PR.

**Review by lgsilva3087 — APPROVED** [review]
Please fix the typos requested by Crash before merging.

**Pull request merged**